### PR TITLE
deserializer should load in batches with only function imports

### PR DESCRIPTION
**Pull request by fwilhe**
Closes SAP/cloud-sdk-backlog#ISSUENUMBER.

### CODE REVIEW

File: .changeset/witty-ducks-design.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+'@sap-cloud-sdk/e2e-tests': patch
+'@sap-cloud-sdk/odata-common': patch
+'@sap-cloud-sdk/odata-v2': patch
+'@sap-cloud-sdk/odata-v4': patch
+---
+
+Deserializer load in batches with only function imports
```
Comment (marikaner) [inline .changeset/witty-ducks-design.md:8]: [req] Please rephrase this to a full sentence. I don't really get what is meant by this.

File: packages/odata-common/src/request-builder/action-function-import-request-builder-base.ts [hunk 2]
```diff
@@ -3,13 +3,15 @@ import { HttpResponse } from '@sap-cloud-sdk/http-client';
...
 import { BatchReference, WithBatchReference } from '../request';
+import { DeSerializers } from '../de-serializers/de-serializers';
 import { MethodRequestBuilder } from './request-builder-base';
 
 /**
  * Create OData request to execute a action or function import.
  * @typeParam ReturnT - Type of the function import return value.
  */
 export abstract class ActionFunctionImportRequestBuilderBase<
+    DeSerializersT extends DeSerializers,
```
Comment (marikaner) [inline packages/odata-common/src/request-builder/action-function-import-request-builder-base.ts:14]: [q] This is actually a breaking change. Maybe we should add this as the last type parameter and make it optional + add a todo to change this in the future and give it a compatibility note.
However, this will make things a bit more complicated, so I am wondering whether this is worth it. I doubt anyone is using this directly, therefore maybe it is fine with a compatibility note.

Comment (marikaner) [reply packages/odata-common/src/request-builder/action-function-import-request-builder-base.ts:14]: [req] Please also add the typeParam to the API Doc

File: packages/odata-common/src/request-builder/batch/batch-request-builder.ts [hunk 3]
```diff
@@ -38,15 +38,25 @@ export class BatchRequestBuilder<
...
-    )?.deSerializers;
+
+    const entityApi = first(Object.values(this.getEntityToApiMap()));
+    const entityDeserializer = entityApi?.deSerializers;
+
+    const operationDeserializer = this.getOperationDeserializer(requests);
+
+    this.deSerializers = entityDeserializer
+      ? entityDeserializer
+      : operationDeserializer;
```
Comment (marikaner) [inline packages/odata-common/src/request-builder/batch/batch-request-builder.ts:59]: [pp] This can be shortened quite a bit and I think would be appropriate here as it makes it slightly more readable in my opinion.

```suggestion
    this.deSerializers = entityApi?.deSerializers || this.getOperationDeserializer(requests);
```

File: packages/odata-common/src/request-builder/batch/batch-request-builder.ts [hunk 4]
```diff
@@ -103,6 +113,31 @@ export class BatchRequestBuilder<
     });
     return request;
   }
+
+  private getOperationDeserializer(
```
Comment (marikaner) [inline packages/odata-common/src/request-builder/batch/batch-request-builder.ts:117]: [req/pp] As this returns (de-)**S**erialiser**S** I would suggest to name it that 😊 Note that those are serializers and deserializers and therefore `deSerializers` not `deserializers`.
```suggestion
  private getOperationDeSerializers(
```

File: packages/odata-common/src/request-builder/batch/batch-request-builder.ts [hunk 5]
```diff
@@ -103,6 +113,31 @@ export class BatchRequestBuilder<
...
+    )[]
+  ) {
+    const outerRequest = first(requests);
+    if (outerRequest instanceof BatchChangeSet) {
+      const innerRequest = first(outerRequest.requests);
+      if (innerRequest && isActionOrFunctionImport(innerRequest)) {
+        return innerRequest._deSerializers;
+      }
+    }
+    return undefined;
```
Comment (marikaner) [inline packages/odata-common/src/request-builder/batch/batch-request-builder.ts:139]: [pp] this line is unnecessary
```suggestion
```

File: test-packages/e2e-tests/test/batch.spec.ts [hunk 6]
```diff
@@ -107,4 +108,43 @@ describe('batch', () => {
     expect(changesetResponse.isSuccess()).toBe(true);
     expect(deleteRes.isSuccess()).toBe(true);
   });
+
+  it('deserializer should load in batches with only function imports', async () => {
```
Comment (marikaner) [inline test-packages/e2e-tests/test/batch.spec.ts:112]:
```suggestion
  it('(de-)serializers should be used in batch requests with only actions', async () => {
```

File: test-packages/e2e-tests/test/batch.spec.ts [hunk 7]
```diff
@@ -107,4 +108,43 @@ describe('batch', () => {
...
+      const deSerializedTestEntity = r.responses[0].as?.(testEntityApi);
+      expect(deSerializedTestEntity).toBeDefined();
+    } else {
+      throw new Error(
+        'Expected response to be of type WriteResponse which is not the case.'
+      );
+    }
+  });
+
+  it('deserializer should load in batches with only bound function imports', async () => {
```
Comment (marikaner) [inline test-packages/e2e-tests/test/batch.spec.ts:129]:
```suggestion
  it('(de-)serializers should be used in batch requests with only bound actions', async () => {
```

File: test-packages/e2e-tests/test/batch.spec.ts [hunk 8]
```diff
@@ -107,4 +108,43 @@ describe('batch', () => {
...
   });
+
+  it('deserializer should load in batches with only function imports', async () => {
+    // See https://github.com/SAP/cloud-sdk-js/issues/3539 for the original issue
+    const results = await batch(changeset(createTestEntityById({ id: 77 })))
+      .withSubRequestPathType('relativeToEntity')
+      .execute(destination);
+
+    const r = results[0];
+    if (r && r.isWriteResponses()) {
```
Comment (marikaner) [inline test-packages/e2e-tests/test/batch.spec.ts:119]: [pp] I am not a fan of variables called `r` (except maybe in arrow functions - which is not the case here)
Also, some of this code can be shortened, so naming gets easier:
```suggestion
    const changesetRes = first(await batch(changeset(createTestEntityById({ id: 77 })))
      .withSubRequestPathType('relativeToEntity')
      .execute(destination))

    if (changesetRes?.isWriteResponses()) {
```

(also applies to the test below)

**Review by marikaner — CHANGES_REQUESTED** [review]
Mostly text and code style, otherwise the solution seems quite legit.

File: packages/odata-common/src/request-builder/batch/batch-request-builder.ts (hunk 4 above)
Comment (fwilhe) [reply packages/odata-common/src/request-builder/batch/batch-request-builder.ts:111]: applied

File: .changeset/witty-ducks-design.md [hunk 9]
```diff
@@ -0,0 +1,8 @@
+---
+'@sap-cloud-sdk/e2e-tests': patch
+'@sap-cloud-sdk/odata-common': patch
+'@sap-cloud-sdk/odata-v2': patch
+'@sap-cloud-sdk/odata-v4': patch
+---
+
+[Compatibility Note] Fix deserializer so it works in batch requests which only contain functions or actions. This changes the type parameters of `ActionFunctionImportRequestBuilderBase`.
```
Comment (marikaner) [inline .changeset/witty-ducks-design.md:8]: [req] We should have two changesets: one for the compatibility note and one for the fix.

Comment (marikaner) [inline .changeset/witty-ducks-design.md:8]: [req] Please note that we are talking about deserializers and serializers, therefore (de-)serializers. Also, as this is a collection of functions we use the plural throughout the docs.

Comment (marikaner) [inline .changeset/witty-ducks-design.md:8]: [req] And last point: We did not change anything in the (de-)serializers, therefore "fix deserializer" does not really describe what happened. My suggestion:

1. [Fixed Issue] Fix batch requests to accept only functions or actions.
2. [Compatibility Note] The `ActionFunctionImportRequestBuilderBase` has an additional mandatory type parameter to pass the according (de-)serializer type. This was missing previously due to a bug.

Of course feel free to chose better words. ;)

File: packages/odata-common/src/request-builder/action-function-import-request-builder-base.ts [hunk 10]
```diff
@@ -3,13 +3,17 @@ import { HttpResponse } from '@sap-cloud-sdk/http-client';
 import { v4 as uuid } from 'uuid';
 import { ODataRequestConfig } from '../request/odata-request-config';
 import { BatchReference, WithBatchReference } from '../request';
+import { DeSerializers } from '../de-serializers/de-serializers';
 import { MethodRequestBuilder } from './request-builder-base';
 
 /**
  * Create OData request to execute a action or function import.
+ * @typeParam DeSerializersT - Type of the (De-)serializers.
```
Comment (marikaner) [inline packages/odata-common/src/request-builder/action-function-import-request-builder-base.ts:11]:
```suggestion
 * @typeParam DeSerializersT - Type of the (de-)serializers.
```

**Review by marikaner — APPROVED** [review]
I had some nitpicky comments on the changeset - the devil lies in the details.
But looks very good otherwise. 👍

File: .changeset/witty-ducks-design.md (hunk 9 above)
Comment (fwilhe) [reply .changeset/witty-ducks-design.md:8]: got it, thanks for the explanation.

Comment (fwilhe) [reply .changeset/witty-ducks-design.md:8]: I think that wording is very good. adapted.
